### PR TITLE
Reworked is_domain_valid

### PR DESCRIPTION
**Pull request by aronmolnar**

### CODE REVIEW

**Comment by al3xdelarge** [comment]
Sorry for the first version. That was bullshit.
Hope this one fits better.

File: w3af/core/data/parsers/doc/url.py [hunk 1]
```diff
@@ -228,23 +235,28 @@ def __init__(self, data, encoding=DEFAULT_ENCODING):
...
+        parsed = urlparse.urlparse(data)
         #
         # This is the case when someone creates a URL like
         # this: URL('www.w3af.com')
         #
-        if scheme == netloc == '' and not path.startswith(u'/'):
+        if parsed.scheme == parsed.netloc == '' and not parsed.path.startswith(u'/'):
             # By default we set the protocol to "http"
             scheme = u'http'
             netloc = path
```
Comment (andresriancho) [inline w3af/core/data/parsers/doc/url.py]: `path` is not defined here.

File: w3af/core/data/parsers/doc/url.py [hunk 2]
```diff
@@ -214,6 +220,7 @@ def __init__(self, data, encoding=DEFAULT_ENCODING):
         self._netloc = None
         self._path = None
         self._params = None
+        self._hostname = None
```
Comment (andresriancho) [inline w3af/core/data/parsers/doc/url.py]: You added this attribute here and to `__slots__` and then never used it? It is confusing to see `self._hostname` and `self.hostname`

File: w3af/core/data/parsers/doc/url.py [hunk 3]
```diff
@@ -560,8 +572,22 @@ def set_domain(self, new_domain):
...
-        """
-        return self.IS_VALID_DOMAIN_RE.match(self.netloc) is not None
+        """        
+        # check if domain name valid
+        if not self.RE_DOMAIN.match(self.hostname):
+            # not a valid domain - maybe an IP address
+            # Check IPv4
+            try:
+                # Check IPv4
+                socket.inet_aton(parsed_url.hostname)
```
Comment (andresriancho) [inline w3af/core/data/parsers/doc/url.py]: `parsed_url` is not defined.

**Review by andresriancho — CHANGES_REQUESTED** [review]
```
[eth:~/tools/w3af] [w3af] al3xdelarge-master+* 130 ± nosetests w3af/core/data/parsers/doc/tests/test_url.py
....................SE...........E..............................EE.............................S..................
======================================================================
ERROR: http is the default protocol, we can provide URLs with no proto
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eth/tools/w3af/w3af/core/data/parsers/doc/tests/test_url.py", line 59, in test_default_proto
    u = URL('w3af.com')
  File "/home/eth/tools/w3af/w3af/core/data/parsers/doc/url.py", line 246, in __init__
    netloc = path
UnboundLocalError: local variable 'path' referenced before assignment

======================================================================
ERROR: test_from_url (w3af.core.data.parsers.doc.tests.test_url.TestURLParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eth/tools/w3af/w3af/core/data/parsers/doc/tests/test_url.py", line 687, in test_from_url
    o = URL('w3af.com')
  File "/home/eth/tools/w3af/w3af/core/data/parsers/doc/url.py", line 246, in __init__
    netloc = path
UnboundLocalError: local variable 'path' referenced before assignment

======================================================================
ERROR: test_is_valid_domain_invalid (w3af.core.data.parsers.doc.tests.test_url.TestURLParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eth/tools/w3af/w3af/core/data/parsers/doc/tests/test_url.py", line 604, in test_is_valid_domain_invalid
    self.assertFalse(URL('http://aaa*a').is_valid_domain())
  File "/home/eth/tools/w3af/w3af/core/data/parsers/doc/url.py", line 582, in is_valid_domain
    socket.inet_aton(parsed_url.hostname)
NameError: global name 'parsed_url' is not defined

======================================================================
ERROR: test_is_valid_domain_valid (w3af.core.data.parsers.doc.tests.test_url.TestURLParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eth/tools/w3af/w3af/core/data/parsers/doc/tests/test_url.py", line 592, in test_is_valid_domain_valid
    self.assertTrue(URL('http://1.2.3.4').is_valid_domain())
  File "/home/eth/tools/w3af/w3af/core/data/parsers/doc/url.py", line 582, in is_valid_domain
    socket.inet_aton(parsed_url.hostname)
NameError: global name 'parsed_url' is not defined

----------------------------------------------------------------------
Ran 114 tests in 0.093s

FAILED (SKIP=2, errors=4)
```

**Comment by andresriancho** [comment]
All tests fail!

```
[eth:~/tools/w3af-experiments] [w3af] al3xdelarge-master* 130 ± nosetests w3af/core/data/parsers/doc/tests/test_url.py
EEEEEEEEEEEEEEEEEEEESEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE.EEEEEEEEE........EEEEEEEEEEEEEEEESEEEEEEEEEEEEEEEEEE
======================================================================
ERROR: test_add (w3af.core.data.parsers.doc.tests.test_url.TestURLParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eth/tools/w3af-experiments/w3af/core/data/parsers/doc/tests/test_url.py", line 997, in test_add
    u = URL('http://www.w3af.com/')
  File "/home/eth/tools/w3af-experiments/w3af/core/data/parsers/doc/url.py", line 256, in __init__
    self.hostname = parsed.hostname or u''
AttributeError: 'URL' object has no attribute 'hostname'

======================================================================
ERROR: test_all_but_scheme (w3af.core.data.parsers.doc.tests.test_url.TestURLParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eth/tools/w3af-experiments/w3af/core/data/parsers/doc/tests/test_url.py", line 547, in test_all_but_scheme
    self.assertEqual(URL('https://w3af.com:443/xyz/').all_but_scheme(),
  File "/home/eth/tools/w3af-experiments/w3af/core/data/parsers/doc/url.py", line 256, in __init__
    self.hostname = parsed.hostname or u''
AttributeError: 'URL' object has no attribute 'hostname'
```

Are you running a test scan after making your changes?

**Comment by al3xdelarge** [comment]
Sorry. Now they should pass.

**Comment by andresriancho** [comment]
Merged!

giphy

**Pull request closed without merging**